lib/decorators/state-config.ts

```diff
@@ -0,0 +1,112 @@
+import {componentStore, bundleStore, providerStore} from '../writers';
+import Module from '../classes/module';
+import {Providers} from './providers';
+import {componentHooks} from "./component";
+import {createConfigErrorMessage} from '../util/helpers';
+import {getInjectableName} from '../util/get-injectable-name';
+import IState = ng.ui.IState;
+import IStateProvider = ng.ui.IStateProvider;
+
+const configsKey = 'ui-router.stateConfigs';
+const childConfigsKey = 'ui-router.stateChildConfigs';
+const annotatedResolvesKey = 'ui-router.annotatedResolves';
+const resolvedMapKey = 'ui-router.resolvedMap';
+
+export interface IComponentState extends IState {
+    component: any;
+}
+
+export function StateConfig(stateConfigs: IComponentState[]){
+    return function(t: any){
+        Providers(...stateConfigs.map(sc => sc.component))(t, `while analyzing StateConfig '${t.name}' state components`);
+        componentStore.set(childConfigsKey, stateConfigs, t);
+        stateConfigs.forEach(config => {
+            if (!config.component) return;
+            let existingConfigs = componentStore.get(configsKey, config.component) || [];
+            componentStore.set(configsKey, [...existingConfigs, config], config.component);
+        });
+    }
+}
+
+function targetIsStaticFn(t) {
+    return t.name !== undefined && t.constructor.name === 'Function';
+}
+
+export function Resolve(resolveName: string = null){
```

Comment:
Yes this function should work just like a regular resolve function so you should be able to return a promise.